### PR TITLE
Fix tokenizing a single-line command chain

### PR DESCRIPTION
**Pull request by aldot**
many thanks in advance && cheers!

### CODE REVIEW

**Review by jordansissel — CHANGES_REQUESTED** [review]
Thanks for all the `free()` changes. I didn't remember that free(NULL) is fine, so thanks for removing all those extra `if (blah != NULL) { free(blah); }` parts!

I added some inline comments. The change to fgets() doesn't seem right, though I don't think it would be harmful.

Can you show an example xdotool invocation that this fixes? I'm not quite understanding the problem being solved and need your help :)

**Comment by aldot** [comment]
On Sat, 13 Feb 2021 at 09:01, Jordan Sissel <notifications@github.com> wrote:

thanks for getting back to me regarding these patchlets!

> I added some inline comments. The change to fgets() doesn't seem right, though I don't think it would be harmful.
>
> Can you show an example xdotool invocation that this fixes? I'm not quite understanding the problem being solved and need your help :)
>
> ________________________________
>
> In xdotool.c:
>
> > @@ -365,6 +363,8 @@ int script_main(int argc, char **argv) {
>    while (fgets(buffer, 4096, input) != NULL) {
>      char *line = buffer;
>      token = NULL;
> +    /* Make sure line is (string) terminated when tokenizing */
> +    buffer[strlen(buffer) + 1] = 0;
>
> This shouldn't be necessary, I think? From glibc's manpage for fgets():
>
> fgets() reads in at most one less than size characters from stream and
> stores them into the buffer pointed to by s. Reading stops after an
> EOF or a newline. If a newline is read, it is stored into the buffer.
> A terminating null byte ('\0') is stored after the last character in
> the buffer.
>
> fgets(buffer, 4096, ...) seems to promise that at most the string will be 4095 bytes long with a null byte following wherever it might be appropriate (at most, byte 4096).

AFAIR the problem comes about the later handling of line while tokenizing:
---8<---
    /* tokenize line into script_argv... */
    while (strlen(line)) {
...
      /* append token */
      if (token != NULL) {
...
      /* advance line to the next token */
      line += strlen(line) + 1; /* !!! points to garbage !!! */
      line += strspn(line, " \t");
    } /* while line being tokenized */
---8<---
Now when there is only one NUL char at the end of buffer we advance
past that and expect the next token to be there.
But what we see there is in fact some random garbage, either from some
previous filling of the buffer, or, if the command was longer than 4k
without an intermittent newline, we escape the buffer and see random
values on the stack past the buffer.
The least intrusive way to fix this, without rewriting the guts of the
function as a whole would IMHO be to make sure we cannot advance past
the actual commands in the buffer by having a second trailing zero
byte that will break out of the 'tokenize line' loop since
strlen(line) would be 0, thanks to that second zero byte.

See?

**Comment by aldot** [comment]
On Sun, 14 Feb 2021 at 09:59, Bernhard Reutner-Fischer
<rep.dot.nop@gmail.com> wrote:
>
> On Sat, 13 Feb 2021 at 09:01, Jordan Sissel <notifications@github.com> wrote:
>
> thanks for getting back to me regarding these patchlets!
>
> > I added some inline comments. The change to fgets() doesn't seem right, though I don't think it would be harmful.
> >
> > Can you show an example xdotool invocation that this fixes? I'm not quite understanding the problem being solved and need your help :)
> >
> > ________________________________
> >
> > In xdotool.c:
> >
> > > @@ -365,6 +363,8 @@ int script_main(int argc, char **argv) {
> >    while (fgets(buffer, 4096, input) != NULL) {
> >      char *line = buffer;
> >      token = NULL;
> > +    /* Make sure line is (string) terminated when tokenizing */
> > +    buffer[strlen(buffer) + 1] = 0;
> >
> > This shouldn't be necessary, I think? From glibc's manpage for fgets():
> >
> > fgets() reads in at most one less than size characters from stream and
> > stores them into the buffer pointed to by s. Reading stops after an
> > EOF or a newline. If a newline is read, it is stored into the buffer.
> > A terminating null byte ('\0') is stored after the last character in
> > the buffer.
> >
> > fgets(buffer, 4096, ...) seems to promise that at most the string will be 4095 bytes long with a null byte following wherever it might be appropriate (at most, byte 4096).
>
> AFAIR the problem comes about the later handling of line while tokenizing:
> ---8<---
>     /* tokenize line into script_argv... */
>     while (strlen(line)) {
> ...
>       /* append token */
>       if (token != NULL) {
> ...
>       /* advance line to the next token */
>       line += strlen(line) + 1; /* !!! points to garbage !!! */
>       line += strspn(line, " \t");
>     } /* while line being tokenized */
> ---8<---
> Now when there is only one NUL char at the end of buffer we advance
> past that and expect the next token to be there.
> But what we see there is in fact some random garbage, either from some
> previous filling of the buffer, or, if the command was longer than 4k
> without an intermittent newline, we escape the buffer and see random
> values on the stack past the buffer.
> The least intrusive way to fix this, without rewriting the guts of the
> function as a whole would IMHO be to make sure we cannot advance past
> the actual commands in the buffer by having a second trailing zero
> byte that will break out of the 'tokenize line' loop since
> strlen(line) would be 0, thanks to that second zero byte.
>
> See?

PS: Note that i did not address the issue that if a token "crosses"
the buffer boundary mid-word then it does not seem to be recognized
properly. Like:
$ printf "mousemove%-4083s--polar -- 0 0 sleep 0.1" " " | /usr/bin/xdotool -
Usage: mousemove [options] <x> <y>
-c, --clearmodifiers      - reset active modifiers (alt, etc) while typing
--screen SCREEN           - which screen to move on, default is current screen
--sync                    - only exit once the mouse has moved
-w, --window <windowid>   - specify a window to move relative to.
You specified the wrong number of args (expected 2 coordinates or 'restore').
/usr/bin/xdotool: Unknown command: olar
Run '/usr/bin/xdotool help' if you want a command list

**Comment by aldot** [comment]
On Sun, 14 Feb 2021 at 10:25, Bernhard Reutner-Fischer
<rep.dot.nop@gmail.com> wrote:
>
> On Sun, 14 Feb 2021 at 09:59, Bernhard Reutner-Fischer
> <rep.dot.nop@gmail.com> wrote:
> >
> > On Sat, 13 Feb 2021 at 09:01, Jordan Sissel <notifications@github.com> wrote:
> >
> > thanks for getting back to me regarding these patchlets!
> >
> > > I added some inline comments. The change to fgets() doesn't seem right, though I don't think it would be harmful.
> > >
> > > Can you show an example xdotool invocation that this fixes? I'm not quite understanding the problem being solved and need your help :)
> > >
> > > ________________________________
> > >
> > > In xdotool.c:
> > >
> > > > @@ -365,6 +363,8 @@ int script_main(int argc, char **argv) {
> > >    while (fgets(buffer, 4096, input) != NULL) {
> > >      char *line = buffer;
> > >      token = NULL;
> > > +    /* Make sure line is (string) terminated when tokenizing */
> > > +    buffer[strlen(buffer) + 1] = 0;
> > >
> > > This shouldn't be necessary, I think? From glibc's manpage for fgets():
> > >
> > > fgets() reads in at most one less than size characters from stream and
> > > stores them into the buffer pointed to by s. Reading stops after an
> > > EOF or a newline. If a newline is read, it is stored into the buffer.
> > > A terminating null byte ('\0') is stored after the last character in
> > > the buffer.
> > >
> > > fgets(buffer, 4096, ...) seems to promise that at most the string will be 4095 bytes long with a null byte following wherever it might be appropriate (at most, byte 4096).
> >
> > AFAIR the problem comes about the later handling of line while tokenizing:
> > ---8<---
> >     /* tokenize line into script_argv... */
> >     while (strlen(line)) {
> > ...
> >       /* append token */
> >       if (token != NULL) {
> > ...
> >       /* advance line to the next token */
> >       line += strlen(line) + 1; /* !!! points to garbage !!! */
> >       line += strspn(line, " \t");
> >     } /* while line being tokenized */
> > ---8<---
> > Now when there is only one NUL char at the end of buffer we advance
> > past that and expect the next token to be there.
> > But what we see there is in fact some random garbage, either from some
> > previous filling of the buffer, or, if the command was longer than 4k
> > without an intermittent newline, we escape the buffer and see random
> > values on the stack past the buffer.
> > The least intrusive way to fix this, without rewriting the guts of the
> > function as a whole would IMHO be to make sure we cannot advance past
> > the actual commands in the buffer by having a second trailing zero
> > byte that will break out of the 'tokenize line' loop since
> > strlen(line) would be 0, thanks to that second zero byte.
> >
> > See?
>
> PS: Note that i did not address the issue that if a token "crosses"
> the buffer boundary mid-word then it does not seem to be recognized
> properly. Like:
> $ printf "mousemove%-4083s--polar -- 0 0 sleep 0.1" " " | /usr/bin/xdotool -
> Usage: mousemove [options] <x> <y>
> -c, --clearmodifiers      - reset active modifiers (alt, etc) while typing
> --screen SCREEN           - which screen to move on, default is current screen
> --sync                    - only exit once the mouse has moved
> -w, --window <windowid>   - specify a window to move relative to.
> You specified the wrong number of args (expected 2 coordinates or 'restore').
> /usr/bin/xdotool: Unknown command: olar
> Run '/usr/bin/xdotool help' if you want a command list

That would be fixed like in the attached sample; See reproducer at the
head of the file, fwiw.
thanks,

**Comment by aldot** [comment]
I have amended the 'Fix tokenizing' patch to also fix tokenizing input that crosses buffer boundaries and adjusted the condition to advance the line buffer so to fix the bug i initially saw in a way that does not increase the buffer size.

File: xdotool.c [hunk 1]
```diff
@@ -365,6 +363,8 @@ int script_main(int argc, char **argv) {
   while (fgets(buffer, 4096, input) != NULL) {
     char *line = buffer;
     token = NULL;
+    /* Make sure line is (string) terminated when tokenizing */
+    buffer[strlen(buffer) + 1] = 0;
```
[the remark this answers is not in the document]
Comment (aldot) [reply xdotool.c]: this was rewritten some time ago

**Pull request closed without merging**